{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Knowledge Graph-based Inference Engine for Thoracic Aortic Aneurysm",
   "id": "49ecbbddc72fc1bc"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-10T03:45:25.044834Z",
     "start_time": "2024-12-10T03:45:14.053341Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import sys         \n",
    "sys.path.insert(0, 'C:/bmin520/python')        \n",
    "from utils import (load_nodes,load_edges,filter_nodes_by_keywords,build_graph,get_nodes_within_distance,subset_graph,save_subgraph)\n",
    "\n",
    "# Set  project directory (This was programmed on a Windows machine)\n",
    "proj_dir = 'C:/bmin520'\n",
    "os.chdir(proj_dir)\n",
    "\n",
    "# Directories\n",
    "primekg_dir = os.path.join(proj_dir, \"PrimeKG/datasets/data\")\n",
    "pmbb_dir = os.path.join(proj_dir, \"PMBB-Release-2020-2.3\")\n",
    "kg_file = os.path.join(primekg_dir, \"kg.csv\")\n",
    "nodes_file = os.path.join(primekg_dir, \"nodes.csv\")\n",
    "icd9_file = os.path.join(pmbb_dir, \"pmbb_icd-9-matrix.txt\")\n",
    "icd10_file = os.path.join(pmbb_dir, \"pmbb_icd-10-matrix.txt\")\n",
    "# variants_file = os.path.join(pmbb_dir, \"pmbb_variants.txt\")\n",
    "\n",
    "# Read and summarize PrimeKG data\n",
    "primekg = pd.read_csv(kg_file, low_memory=False)\n",
    "primekg.query('y_type==\"disease\"|x_type==\"disease\"')"
   ],
   "id": "initial_id",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                 relation  display_relation  x_index                 x_id  \\\n",
       "346728   contraindication  contraindication    15193              DB05271   \n",
       "346729   contraindication  contraindication    15193              DB05271   \n",
       "346730         indication        indication    16687              DB00492   \n",
       "346731         indication        indication    16687              DB00492   \n",
       "346732   contraindication  contraindication    14483              DB13956   \n",
       "...                   ...               ...      ...                  ...   \n",
       "6499510  exposure_disease         linked to    31393                11565   \n",
       "6499511  exposure_disease         linked to    28208                 5148   \n",
       "6499512  exposure_disease         linked to    84172                 1076   \n",
       "6499513  exposure_disease         linked to    33063               100130   \n",
       "6499514  exposure_disease         linked to    28668  9976_4580_1451_1455   \n",
       "\n",
       "          x_type                                     x_name       x_source  \\\n",
       "346728      drug                                 Rotigotine       DrugBank   \n",
       "346729      drug                                 Rotigotine       DrugBank   \n",
       "346730      drug                                 Fosinopril       DrugBank   \n",
       "346731      drug                                 Fosinopril       DrugBank   \n",
       "346732      drug                         Estradiol valerate       DrugBank   \n",
       "...          ...                                        ...            ...   \n",
       "6499510  disease                       metabolic syndrome X          MONDO   \n",
       "6499511  disease                   type 2 diabetes mellitus          MONDO   \n",
       "6499512  disease                        glucose intolerance          MONDO   \n",
       "6499513  disease  adult acute respiratory distress syndrome          MONDO   \n",
       "6499514  disease                       retinal degeneration  MONDO_grouped   \n",
       "\n",
       "         y_index                         y_id    y_type  \\\n",
       "346728     33577                         5044   disease   \n",
       "346729     36035  1200_1134_15512_5080_100078   disease   \n",
       "346730     33577                         5044   disease   \n",
       "346731     36035  1200_1134_15512_5080_100078   disease   \n",
       "346732     33577                         5044   disease   \n",
       "...          ...                          ...       ...   \n",
       "6499510    61825                      D015032  exposure   \n",
       "6499511    61825                      D015032  exposure   \n",
       "6499512    61825                      D015032  exposure   \n",
       "6499513    61825                      D015032  exposure   \n",
       "6499514    62183                      D015039  exposure   \n",
       "\n",
       "                        y_name       y_source  \n",
       "346728   hypertensive disorder          MONDO  \n",
       "346729            hypertension  MONDO_grouped  \n",
       "346730   hypertensive disorder          MONDO  \n",
       "346731            hypertension  MONDO_grouped  \n",
       "346732   hypertensive disorder          MONDO  \n",
       "...                        ...            ...  \n",
       "6499510                   Zinc            CTD  \n",
       "6499511                   Zinc            CTD  \n",
       "6499512                   Zinc            CTD  \n",
       "6499513                   Zinc            CTD  \n",
       "6499514                  Ziram            CTD  \n",
       "\n",
       "[618100 rows x 12 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation</th>\n",
       "      <th>display_relation</th>\n",
       "      <th>x_index</th>\n",
       "      <th>x_id</th>\n",
       "      <th>x_type</th>\n",
       "      <th>x_name</th>\n",
       "      <th>x_source</th>\n",
       "      <th>y_index</th>\n",
       "      <th>y_id</th>\n",
       "      <th>y_type</th>\n",
       "      <th>y_name</th>\n",
       "      <th>y_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>346728</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>15193</td>\n",
       "      <td>DB05271</td>\n",
       "      <td>drug</td>\n",
       "      <td>Rotigotine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>33577</td>\n",
       "      <td>5044</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertensive disorder</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346729</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>15193</td>\n",
       "      <td>DB05271</td>\n",
       "      <td>drug</td>\n",
       "      <td>Rotigotine</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>36035</td>\n",
       "      <td>1200_1134_15512_5080_100078</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertension</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346730</th>\n",
       "      <td>indication</td>\n",
       "      <td>indication</td>\n",
       "      <td>16687</td>\n",
       "      <td>DB00492</td>\n",
       "      <td>drug</td>\n",
       "      <td>Fosinopril</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>33577</td>\n",
       "      <td>5044</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertensive disorder</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346731</th>\n",
       "      <td>indication</td>\n",
       "      <td>indication</td>\n",
       "      <td>16687</td>\n",
       "      <td>DB00492</td>\n",
       "      <td>drug</td>\n",
       "      <td>Fosinopril</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>36035</td>\n",
       "      <td>1200_1134_15512_5080_100078</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertension</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346732</th>\n",
       "      <td>contraindication</td>\n",
       "      <td>contraindication</td>\n",
       "      <td>14483</td>\n",
       "      <td>DB13956</td>\n",
       "      <td>drug</td>\n",
       "      <td>Estradiol valerate</td>\n",
       "      <td>DrugBank</td>\n",
       "      <td>33577</td>\n",
       "      <td>5044</td>\n",
       "      <td>disease</td>\n",
       "      <td>hypertensive disorder</td>\n",
       "      <td>MONDO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499510</th>\n",
       "      <td>exposure_disease</td>\n",
       "      <td>linked to</td>\n",
       "      <td>31393</td>\n",
       "      <td>11565</td>\n",
       "      <td>disease</td>\n",
       "      <td>metabolic syndrome X</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>61825</td>\n",
       "      <td>D015032</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Zinc</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499511</th>\n",
       "      <td>exposure_disease</td>\n",
       "      <td>linked to</td>\n",
       "      <td>28208</td>\n",
       "      <td>5148</td>\n",
       "      <td>disease</td>\n",
       "      <td>type 2 diabetes mellitus</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>61825</td>\n",
       "      <td>D015032</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Zinc</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499512</th>\n",
       "      <td>exposure_disease</td>\n",
       "      <td>linked to</td>\n",
       "      <td>84172</td>\n",
       "      <td>1076</td>\n",
       "      <td>disease</td>\n",
       "      <td>glucose intolerance</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>61825</td>\n",
       "      <td>D015032</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Zinc</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499513</th>\n",
       "      <td>exposure_disease</td>\n",
       "      <td>linked to</td>\n",
       "      <td>33063</td>\n",
       "      <td>100130</td>\n",
       "      <td>disease</td>\n",
       "      <td>adult acute respiratory distress syndrome</td>\n",
       "      <td>MONDO</td>\n",
       "      <td>61825</td>\n",
       "      <td>D015032</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Zinc</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6499514</th>\n",
       "      <td>exposure_disease</td>\n",
       "      <td>linked to</td>\n",
       "      <td>28668</td>\n",
       "      <td>9976_4580_1451_1455</td>\n",
       "      <td>disease</td>\n",
       "      <td>retinal degeneration</td>\n",
       "      <td>MONDO_grouped</td>\n",
       "      <td>62183</td>\n",
       "      <td>D015039</td>\n",
       "      <td>exposure</td>\n",
       "      <td>Ziram</td>\n",
       "      <td>CTD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>618100 rows Ã— 12 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Subset KG to thoracic aortic aneurysm-related nodes and edges",
   "id": "20b19486d03e9f98"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-10T01:25:53.694152Z",
     "start_time": "2024-12-10T01:25:45.486867Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Check if the subgraph directory exists and create it if not\n",
    "subgraph_data_dir = os.path.join(proj_dir, \"subgraph\")\n",
    "os.makedirs(subgraph_data_dir, exist_ok=True)\n",
    "\n",
    "nodes = load_nodes(primekg_dir)\n",
    "edges = load_edges(primekg_dir)\n",
    "g = build_graph(nodes, edges)\n",
    "\n",
    "# keywords to subset nodes by (courtesy of Teddy)\n",
    "keywords = [\"aortic aneurysm, familial thoracic\", \"thoracic aortic aneurysm\", \n",
    "            \"aortic aneurysm, thoracic\", \"aneurysm of sinus of Valsalva\", \n",
    "            \"aortic aneurysm (disease)\", \"aneurysm or dilation of ascending aorta\",\n",
    "            \"congenital aneurysms of the great vessels\", \"Thoracic aortic aneurysm\", \n",
    "            \"Ascending tubular aorta aneurysm\", \"Descending thoracic aorta aneurysm\",\n",
    "            \"Aortic arch aneurysm\", \"Aortic root aneurysm\",\n",
    "            \"Fusiform descending thoracic aortic aneurysm\", \"Saccular descending thoracic aortic aneurysm\",\n",
    "            \"Fusiform ascending tubular aorta aneurysm\", \"Fusiform aortic arch aneurysm\", \n",
    "            \"Saccular aortic arch aneurysm\"]\n",
    "\n",
    "keyword_nodes = filter_nodes_by_keywords(nodes, keywords)\n",
    "\n",
    "# Get nodes within max_distance from the keyword-related nodes using BFS\n",
    "start_nodes = keyword_nodes['node_index'].tolist()\n",
    "nodes_to_keep = get_nodes_within_distance(g, start_nodes, max_distance=1)\n",
    "\n",
    "# Subset graph and save\n",
    "sub_nodes, sub_edges = subset_graph(nodes, edges, nodes_to_keep)\n",
    "out_dir = subgraph_data_dir\n",
    "save_subgraph(sub_nodes, sub_edges, out_dir)\n",
    "\n",
    "print(\"Subgraph creation complete. Files saved.\")"
   ],
   "id": "b9b105c673e3e",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\bmin520/python\\utils.py:24: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.\n",
      "  mask = nodes['node_name'].str.contains('|'.join(keywords), case=False, na=False)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subgraph creation complete. Files saved.\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-10T03:04:45.853123Z",
     "start_time": "2024-12-10T03:04:44.977194Z"
    }
   },
   "cell_type": "code",
   "source": [
    "drug_features_file = os.path.join(primekg_dir, \"drug_features.csv\")\n",
    "disease_features_file = os.path.join(primekg_dir, \"disease_features.csv\")\n",
    "\n",
    "# Load feature files and subset\n",
    "drug_features = pd.read_csv(drug_features_file)\n",
    "disease_features = pd.read_csv(disease_features_file)\n",
    "subset_drug_features = drug_features[drug_features['node_index'].isin(nodes_to_keep)]\n",
    "subset_disease_features = disease_features[disease_features['node_index'].isin(nodes_to_keep)]\n",
    "\n",
    "# Save subsetted features\n",
    "subset_drug_features_file = os.path.join(subgraph_data_dir, \"subset_drug_features.csv\")\n",
    "subset_disease_features_file = os.path.join(subgraph_data_dir, \"subset_disease_features.csv\")\n",
    "subset_drug_features.to_csv(subset_drug_features_file, index=False)\n",
    "subset_disease_features.to_csv(subset_disease_features_file, index=False)"
   ],
   "id": "f3cfa3cba9734ac6",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-09T03:17:13.575237Z",
     "start_time": "2024-12-09T03:17:13.507343Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "data_dir = os.path.join(proj_dir, \"PrimeKG/datasets/data\")\n",
    "\n",
    "# List of key files and description (from PrimeKG repository)\n",
    "files_to_check = [\n",
    "    (\"nodes.csv\", \"Node-level information.\"),\n",
    "    (\"edges.csv\", \"Edges representing undirected relationships between nodes.\"),\n",
    "    (\"kg.csv\", \"The Precision Medicine knowledge graph (joined nodes and edges).\"),\n",
    "    (\"disease_features.csv\", \"Textual descriptions of diseases.\"),\n",
    "    (\"drug_features.csv\", \"Textual descriptions of drugs.\"),\n",
    "    (\"kg_raw.csv\", \"Intermediate KG from joining nodes and edges.\"),\n",
    "    (\"kg_giant.csv\", \"Intermediate KG after taking largest connected component.\"),\n",
    "    (\"kg_grouped.csv\", \"Intermediate KG after grouping diseases.\"),\n",
    "    (\"kg_grouped_diseases.csv\", \"List of all diseases and their assigned group name.\"),\n",
    "    (\"kg_grouped_diseases_bert_map.csv\", \"Manual grouping for diseases using BERT model.\")\n",
    "]18\n",
    "\n",
    "for fname, desc in files_to_check:\n",
    "    fpath = os.path.join(data_dir, fname)\n",
    "    if os.path.exists(fpath):\n",
    "        df = pd.read_csv(fpath, nrows=3)  # just read a few rows\n",
    "        print(f\"File: {fname}\")\n",
    "        print(f\"Description: {desc}\")\n",
    "        print(\"Columns:\", df.columns.tolist())\n",
    "        print(\"First few rows:\")\n",
    "        if len(df.head()) > 100:\n",
    "            short = df.head()[:100] + '...'\n",
    "            print(short, \"\\n\")\n",
    "\n",
    "        else:\n",
    "            print(df.head(), \"\\n\")\n",
    "    else:\n",
    "        print(f\"File not found: {fname}\\n\")\n"
   ],
   "id": "11bb43200e88b411",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File: nodes.csv\n",
      "Description: Node-level information.\n",
      "Columns: ['node_index', 'node_id', 'node_type', 'node_name', 'node_source']\n",
      "First few rows:\n",
      "   node_index  node_id     node_type node_name node_source\n",
      "0           0     9796  gene/protein    PHYHIP        NCBI\n",
      "1           1     7918  gene/protein    GPANK1        NCBI\n",
      "2           2     8233  gene/protein     ZRSR2        NCBI \n",
      "\n",
      "File: edges.csv\n",
      "Description: Edges representing undirected relationships between nodes.\n",
      "Columns: ['relation', 'display_relation', 'x_index', 'y_index']\n",
      "First few rows:\n",
      "          relation display_relation  x_index  y_index\n",
      "0  protein_protein              ppi        0     8889\n",
      "1  protein_protein              ppi        1     2798\n",
      "2  protein_protein              ppi        2     5646 \n",
      "\n",
      "File: kg.csv\n",
      "Description: The Precision Medicine knowledge graph (joined nodes and edges).\n",
      "Columns: ['relation', 'display_relation', 'x_index', 'x_id', 'x_type', 'x_name', 'x_source', 'y_index', 'y_id', 'y_type', 'y_name', 'y_source']\n",
      "First few rows:\n",
      "          relation display_relation  x_index  x_id        x_type  x_name  \\\n",
      "0  protein_protein              ppi        0  9796  gene/protein  PHYHIP   \n",
      "1  protein_protein              ppi        1  7918  gene/protein  GPANK1   \n",
      "2  protein_protein              ppi        2  8233  gene/protein   ZRSR2   \n",
      "\n",
      "  x_source  y_index   y_id        y_type y_name y_source  \n",
      "0     NCBI     8889  56992  gene/protein  KIF15     NCBI  \n",
      "1     NCBI     2798   9240  gene/protein  PNMA1     NCBI  \n",
      "2     NCBI     5646  23548  gene/protein  TTC33     NCBI   \n",
      "\n",
      "File: disease_features.csv\n",
      "Description: Textual descriptions of diseases.\n",
      "Columns: ['node_index', 'mondo_id', 'mondo_name', 'group_id_bert', 'group_name_bert', 'mondo_definition', 'umls_description', 'orphanet_definition', 'orphanet_prevalence', 'orphanet_epidemiology', 'orphanet_clinical_description', 'orphanet_management_and_treatment', 'mayo_symptoms', 'mayo_causes', 'mayo_risk_factors', 'mayo_complications', 'mayo_prevention', 'mayo_see_doc']\n",
      "First few rows:\n",
      "   node_index  mondo_id                                         mondo_name  \\\n",
      "0       27165      8019             mullerian aplasia and hyperandrogenism   \n",
      "1       27165      8019             mullerian aplasia and hyperandrogenism   \n",
      "2       27166     11043  myelodysplasia, immunodeficiency, facial dysmo...   \n",
      "\n",
      "   group_id_bert  group_name_bert  \\\n",
      "0            NaN              NaN   \n",
      "1            NaN              NaN   \n",
      "2            NaN              NaN   \n",
      "\n",
      "                                    mondo_definition  \\\n",
      "0  Deficiency of the glycoprotein WNT4, associate...   \n",
      "1  Deficiency of the glycoprotein WNT4, associate...   \n",
      "2                                                NaN   \n",
      "\n",
      "                                    umls_description  \\\n",
      "0  Deficiency of the glycoprotein wnt4, associate...   \n",
      "1  Deficiency of the glycoprotein wnt4, associate...   \n",
      "2                                                NaN   \n",
      "\n",
      "                                 orphanet_definition  orphanet_prevalence  \\\n",
      "0  A rare syndrome with 46,XX disorder of sex dev...                  NaN   \n",
      "1  A rare syndrome with 46,XX disorder of sex dev...                  NaN   \n",
      "2                                                NaN                  NaN   \n",
      "\n",
      "   orphanet_epidemiology  orphanet_clinical_description  \\\n",
      "0                    NaN                            NaN   \n",
      "1                    NaN                            NaN   \n",
      "2                    NaN                            NaN   \n",
      "\n",
      "   orphanet_management_and_treatment  mayo_symptoms  mayo_causes  \\\n",
      "0                                NaN            NaN          NaN   \n",
      "1                                NaN            NaN          NaN   \n",
      "2                                NaN            NaN          NaN   \n",
      "\n",
      "   mayo_risk_factors  mayo_complications  mayo_prevention  mayo_see_doc  \n",
      "0                NaN                 NaN              NaN           NaN  \n",
      "1                NaN                 NaN              NaN           NaN  \n",
      "2                NaN                 NaN              NaN           NaN   \n",
      "\n",
      "File: drug_features.csv\n",
      "Description: Textual descriptions of drugs.\n",
      "Columns: ['node_index', 'description', 'half_life', 'indication', 'mechanism_of_action', 'protein_binding', 'pharmacodynamics', 'state', 'atc_1', 'atc_2', 'atc_3', 'atc_4', 'category', 'group', 'pathway', 'molecular_weight', 'tpsa', 'clogp']\n",
      "First few rows:\n",
      "   node_index                                        description  \\\n",
      "0       14012  Copper is a transition metal and a trace eleme...   \n",
      "1       14013  Oxygen is an element displayed by the symbol O...   \n",
      "2       14014  Flunisolide (marketed as AeroBid, Nasalide, Na...   \n",
      "\n",
      "                                       half_life  \\\n",
      "0                                            NaN   \n",
      "1  The half-life is approximately 122.24 seconds   \n",
      "2                     The half-life is 1.8 hours   \n",
      "\n",
      "                                          indication  \\\n",
      "0  For use in the supplementation of total parent...   \n",
      "1  Oxygen therapy in clinical settings is used ac...   \n",
      "2  For the maintenance treatment of asthma as a p...   \n",
      "\n",
      "                                 mechanism_of_action  \\\n",
      "0  Copper is absorbed from the gut via high affin...   \n",
      "1  Oxygen therapy increases the arterial pressure...   \n",
      "2  Flunisolide is a glucocorticoid receptor agoni...   \n",
      "\n",
      "                                     protein_binding  \\\n",
      "0  Copper is nearly entirely bound by ceruloplasm...   \n",
      "1  Oxygen binds to oxygen-carrying protein in red...   \n",
      "2            Approximately 40% after oral inhalation   \n",
      "\n",
      "                                    pharmacodynamics                    state  \\\n",
      "0  Copper is incorporated into many enzymes throu...       Copper is a solid.   \n",
      "1  Oxygen therapy improves effective cellular oxy...         Oxygen is a gas.   \n",
      "2  Flunisolide is a synthetic corticosteroid. It ...  Flunisolide is a solid.   \n",
      "\n",
      "                                               atc_1  \\\n",
      "0                                                NaN   \n",
      "1         Oxygen is anatomically related to various.   \n",
      "2  Flunisolide is anatomically related to respira...   \n",
      "\n",
      "                                               atc_2  \\\n",
      "0                                                NaN   \n",
      "1  Oxygen is in the therapeutic group of all othe...   \n",
      "2  Flunisolide is in the therapeutic group of nas...   \n",
      "\n",
      "                                               atc_3  \\\n",
      "0                                                NaN   \n",
      "1  Oxygen is pharmacologically related to all oth...   \n",
      "2  Flunisolide is pharmacologically related to de...   \n",
      "\n",
      "                                               atc_4  \\\n",
      "0                                                NaN   \n",
      "1  The chemical and functional group of  is medic...   \n",
      "2  The chemical and functional group of  is corti...   \n",
      "\n",
      "                                            category  \\\n",
      "0  Copper is part of Copper-containing Intrauteri...   \n",
      "1  Oxygen is part of Chalcogens ; Elements ; Gase...   \n",
      "2  Flunisolide is part of Adrenal Cortex Hormones...   \n",
      "\n",
      "                                          group  pathway  \\\n",
      "0       Copper is approved and investigational.      NaN   \n",
      "1          Oxygen is approved and vet_approved.      NaN   \n",
      "2  Flunisolide is approved and investigational.      NaN   \n",
      "\n",
      "                 molecular_weight  \\\n",
      "0                             NaN   \n",
      "1   The molecular weight is 32.0.   \n",
      "2  The molecular weight is 434.5.   \n",
      "\n",
      "                                                tpsa  \\\n",
      "0                                                NaN   \n",
      "1  Oxygen has a topological polar surface area of...   \n",
      "2  Flunisolide has a topological polar surface ar...   \n",
      "\n",
      "                          clogp  \n",
      "0                           NaN  \n",
      "1                           NaN  \n",
      "2  The log p value of  is 2.41.   \n",
      "\n",
      "File: kg_raw.csv\n",
      "Description: Intermediate KG from joining nodes and edges.\n",
      "Columns: ['relation', 'display_relation', 'x_id', 'x_type', 'x_name', 'x_source', 'y_id', 'y_type', 'y_name', 'y_source']\n",
      "First few rows:\n",
      "          relation display_relation  x_id        x_type  x_name x_source  \\\n",
      "0  protein_protein              ppi  9796  gene/protein  PHYHIP     NCBI   \n",
      "1  protein_protein              ppi  7918  gene/protein  GPANK1     NCBI   \n",
      "2  protein_protein              ppi  8233  gene/protein   ZRSR2     NCBI   \n",
      "\n",
      "    y_id        y_type y_name y_source  \n",
      "0  56992  gene/protein  KIF15     NCBI  \n",
      "1   9240  gene/protein  PNMA1     NCBI  \n",
      "2  23548  gene/protein  TTC33     NCBI   \n",
      "\n",
      "File: kg_giant.csv\n",
      "Description: Intermediate KG after taking largest connected component.\n",
      "Columns: ['relation', 'display_relation', 'x_id', 'x_type', 'x_name', 'x_source', 'y_id', 'y_type', 'y_name', 'y_source']\n",
      "First few rows:\n",
      "          relation display_relation  x_id        x_type  x_name x_source  \\\n",
      "0  protein_protein              ppi  9796  gene/protein  PHYHIP     NCBI   \n",
      "1  protein_protein              ppi  7918  gene/protein  GPANK1     NCBI   \n",
      "2  protein_protein              ppi  8233  gene/protein   ZRSR2     NCBI   \n",
      "\n",
      "    y_id        y_type y_name y_source  \n",
      "0  56992  gene/protein  KIF15     NCBI  \n",
      "1   9240  gene/protein  PNMA1     NCBI  \n",
      "2  23548  gene/protein  TTC33     NCBI   \n",
      "\n",
      "File: kg_grouped.csv\n",
      "Description: Intermediate KG after grouping diseases.\n",
      "Columns: ['relation', 'display_relation', 'x_id', 'x_type', 'x_name', 'x_source', 'y_id', 'y_type', 'y_name', 'y_source']\n",
      "First few rows:\n",
      "          relation display_relation  x_id        x_type  x_name x_source  \\\n",
      "0  protein_protein              ppi  9796  gene/protein  PHYHIP     NCBI   \n",
      "1  protein_protein              ppi  7918  gene/protein  GPANK1     NCBI   \n",
      "2  protein_protein              ppi  8233  gene/protein   ZRSR2     NCBI   \n",
      "\n",
      "    y_id        y_type y_name y_source  \n",
      "0  56992  gene/protein  KIF15     NCBI  \n",
      "1   9240  gene/protein  PNMA1     NCBI  \n",
      "2  23548  gene/protein  TTC33     NCBI   \n",
      "\n",
      "File: kg_grouped_diseases.csv\n",
      "Description: List of all diseases and their assigned group name.\n",
      "Columns: ['node_id', 'node_type', 'node_name', 'node_source', 'group_name_auto', 'group_name_bert']\n",
      "First few rows:\n",
      "   node_id node_type                                          node_name  \\\n",
      "0    13924   disease                    osteogenesis imperfecta type 13   \n",
      "1    11160   disease       autosomal recessive nonsyndromic deafness 15   \n",
      "2     8099   disease  congenital stationary night blindness autosoma...   \n",
      "\n",
      "  node_source                                    group_name_auto  \\\n",
      "0       MONDO                            osteogenesis imperfecta   \n",
      "1       MONDO          autosomal recessive nonsyndromic deafness   \n",
      "2       MONDO  congenital stationary night blindness autosoma...   \n",
      "\n",
      "                                     group_name_bert  \n",
      "0                            osteogenesis imperfecta  \n",
      "1          autosomal recessive nonsyndromic deafness  \n",
      "2  congenital stationary night blindness autosoma...   \n",
      "\n",
      "File: kg_grouped_diseases_bert_map.csv\n",
      "Description: Manual grouping for diseases using BERT model.\n",
      "Columns: ['node_id', 'node_type', 'node_name', 'node_source', 'group_name_auto', 'group_name_bert', 'group_id_bert']\n",
      "First few rows:\n",
      "   node_id node_type                        node_name node_source  \\\n",
      "0    13924   disease  osteogenesis imperfecta type 13       MONDO   \n",
      "1    12592   disease  osteogenesis imperfecta type 11       MONDO   \n",
      "2    14672   disease  osteogenesis imperfecta type 17       MONDO   \n",
      "\n",
      "           group_name_auto          group_name_bert  \\\n",
      "0  osteogenesis imperfecta  osteogenesis imperfecta   \n",
      "1  osteogenesis imperfecta  osteogenesis imperfecta   \n",
      "2  osteogenesis imperfecta  osteogenesis imperfecta   \n",
      "\n",
      "                                       group_id_bert  \n",
      "0  13924_12592_14672_13460_12591_12536_30861_8146...  \n",
      "1  13924_12592_14672_13460_12591_12536_30861_8146...  \n",
      "2  13924_12592_14672_13460_12591_12536_30861_8146...   \n",
      "\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-11T04:20:36.855231Z",
     "start_time": "2024-12-11T04:16:28.485221Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import sys\n",
    "sys.path.insert(0, 'C:/bmin520/python')\n",
    "from utils import filter_pmbb_by_icd, enforce_uniform_sampling\n",
    "import pandas as pd\n",
    "\n",
    "pmbb_dir = \"C:/bmin520/PMBB-Release-2020-2.3\"\n",
    "output_dir = \"C:/bmin520/subgraph/patients_filtered_new\"\n",
    "icd9_code = '441.2'\n",
    "icd10_code = 'I71.2' #TAA, not all have icd9, most have icd10\n",
    "num_patients = 1842\n",
    "seed = 1\n",
    "\n",
    "sampled_ids_file = filter_pmbb_by_icd(icd9_code, icd10_code, pmbb_dir, output_dir, num_patients, seed)\n",
    "\n",
    "sampled_ids = pd.read_csv(sampled_ids_file)['PMBB_ID'].tolist()\n",
    "enforce_uniform_sampling(pmbb_dir, sampled_ids, output_dir)\n"
   ],
   "id": "a2742c2c4d378eed",
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[1;32mIn[2], line 13\u001B[0m\n\u001B[0;32m     10\u001B[0m num_patients \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m1842\u001B[39m\n\u001B[0;32m     11\u001B[0m seed \u001B[38;5;241m=\u001B[39m \u001B[38;5;241m1\u001B[39m\n\u001B[1;32m---> 13\u001B[0m sampled_ids_file \u001B[38;5;241m=\u001B[39m filter_pmbb_by_icd(icd9_code, icd10_code, pmbb_dir, output_dir, num_patients, seed)\n\u001B[0;32m     15\u001B[0m sampled_ids \u001B[38;5;241m=\u001B[39m pd\u001B[38;5;241m.\u001B[39mread_csv(sampled_ids_file)[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mPMBB_ID\u001B[39m\u001B[38;5;124m'\u001B[39m]\u001B[38;5;241m.\u001B[39mtolist()\n\u001B[0;32m     16\u001B[0m enforce_uniform_sampling(pmbb_dir, sampled_ids, output_dir)\n",
      "File \u001B[1;32mC:\\bmin520/python\\utils.py:490\u001B[0m, in \u001B[0;36mfilter_pmbb_by_icd\u001B[1;34m(icd9, icd10, pmbb_dir, output_dir, num_patients, seed)\u001B[0m\n\u001B[0;32m    488\u001B[0m #     icd10_file = os.path.join(pmbb_dir, 'PMBB-Release-2020-2.3_phenotype_icd-10-matrix.txt')\n\u001B[0;32m    489\u001B[0m #     icd9_df = pd.read_csv(icd9_file, sep='\\t', low_memory=False)\n\u001B[1;32m--> 490\u001B[0m #     icd10_df = pd.read_csv(icd10_file, sep='\\t', low_memory=False)\n\u001B[0;32m    491\u001B[0m #\n\u001B[0;32m    492\u001B[0m #     icd10_df.columns = icd10_df.columns.astype(str)  # Ensure ICD-10 codes are strings\n",
      "File \u001B[1;32m~\\anaconda3\\envs\\bmin520\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1026\u001B[0m, in \u001B[0;36mread_csv\u001B[1;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001B[0m\n\u001B[0;32m   1013\u001B[0m kwds_defaults \u001B[38;5;241m=\u001B[39m _refine_defaults_read(\n\u001B[0;32m   1014\u001B[0m     dialect,\n\u001B[0;32m   1015\u001B[0m     delimiter,\n\u001B[1;32m   (...)\u001B[0m\n\u001B[0;32m   1022\u001B[0m     dtype_backend\u001B[38;5;241m=\u001B[39mdtype_backend,\n\u001B[0;32m   1023\u001B[0m )\n\u001B[0;32m   1024\u001B[0m kwds\u001B[38;5;241m.\u001B[39mupdate(kwds_defaults)\n\u001B[1;32m-> 1026\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m _read(filepath_or_buffer, kwds)\n",
      "File \u001B[1;32m~\\anaconda3\\envs\\bmin520\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:626\u001B[0m, in \u001B[0;36m_read\u001B[1;34m(filepath_or_buffer, kwds)\u001B[0m\n\u001B[0;32m    623\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m parser\n\u001B[0;32m    625\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m parser:\n\u001B[1;32m--> 626\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m parser\u001B[38;5;241m.\u001B[39mread(nrows)\n",
      "File \u001B[1;32m~\\anaconda3\\envs\\bmin520\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1923\u001B[0m, in \u001B[0;36mTextFileReader.read\u001B[1;34m(self, nrows)\u001B[0m\n\u001B[0;32m   1916\u001B[0m nrows \u001B[38;5;241m=\u001B[39m validate_integer(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mnrows\u001B[39m\u001B[38;5;124m\"\u001B[39m, nrows)\n\u001B[0;32m   1917\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m   1918\u001B[0m     \u001B[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001B[39;00m\n\u001B[0;32m   1919\u001B[0m     (\n\u001B[0;32m   1920\u001B[0m         index,\n\u001B[0;32m   1921\u001B[0m         columns,\n\u001B[0;32m   1922\u001B[0m         col_dict,\n\u001B[1;32m-> 1923\u001B[0m     ) \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_engine\u001B[38;5;241m.\u001B[39mread(  \u001B[38;5;66;03m# type: ignore[attr-defined]\u001B[39;00m\n\u001B[0;32m   1924\u001B[0m         nrows\n\u001B[0;32m   1925\u001B[0m     )\n\u001B[0;32m   1926\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mException\u001B[39;00m:\n\u001B[0;32m   1927\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mclose()\n",
      "File \u001B[1;32m~\\anaconda3\\envs\\bmin520\\Lib\\site-packages\\pandas\\io\\parsers\\c_parser_wrapper.py:239\u001B[0m, in \u001B[0;36mCParserWrapper.read\u001B[1;34m(self, nrows)\u001B[0m\n\u001B[0;32m    236\u001B[0m         data \u001B[38;5;241m=\u001B[39m _concatenate_chunks(chunks)\n\u001B[0;32m    238\u001B[0m     \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m--> 239\u001B[0m         data \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_reader\u001B[38;5;241m.\u001B[39mread(nrows)\n\u001B[0;32m    240\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mStopIteration\u001B[39;00m:\n\u001B[0;32m    241\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_first_chunk:\n",
      "\u001B[1;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-12-11T04:26:42.131037Z",
     "start_time": "2024-12-11T04:22:09.061529Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Filter icd10 matrix file by sampled patient IDs file and return and save the filtered matrix\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Paths\n",
    "pmbb_dir = \"C:/bmin520/PMBB-Release-2020-2.3\"\n",
    "output_dir = \"C:/bmin520/subgraph/patients_filtered_new\"\n",
    "sampled_ids_file = os.path.join(output_dir, \"sampled_patient_ids.csv\")\n",
    "icd10_file = os.path.join(pmbb_dir, \"PMBB-Release-2020-2.3_phenotype_icd-10-matrix.txt\")\n",
    "\n",
    "sampled_ids = pd.read_csv(sampled_ids_file)['PMBB_ID'].tolist()\n",
    "\n",
    "# Filter ICD-10 matrix\n",
    "icd10_data = pd.read_csv(icd10_file, sep='\\t')\n",
    "filtered_icd10 = icd10_data[icd10_data['PMBB_ID'].isin(sampled_ids)]\n",
    "\n",
    "# Save\n",
    "output_file = os.path.join(output_dir, \"filtered_icd10_matrix.csv\")\n",
    "filtered_icd10.to_csv(output_file, index=False)\n"
   ],
   "id": "cc94d2367b30ccfa",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Come back to later, currently not useful.\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "from pyplink import PyPlink\n",
    "\n",
    "# Paths\n",
    "plink_file_prefix = \"C:/bmin520/PMBB-Release-2020-2.0_genetic_exome_GL\"\n",
    "sampled_ids_file = \"C:/bmin520/subgraph/patients_filtered/sampled_patient_ids.csv\"\n",
    "\n",
    "aortopathy_genes = [\"ACTA2\", \"ADAMTS10\", \"BGN\", \"CBS\", \"COL3A1\",\n",
    "                    \"COL5A1\", \"COL5A2\", \"EFEMP2\",\"FBN1\", \"FBN2\",\n",
    "                    \"FLNA\", \"FOXE3\", \"LOX\", \"MED12\"]\n",
    "\n",
    "sampled_ids = pd.read_csv(sampled_ids_file)['PMBB_ID'].tolist()\n",
    "\n",
    "# Load Plink data\n",
    "pyp = PyPlink(plink_file_prefix)\n",
    "fam_data = pyp.get_fam()\n",
    "filtered_fam = fam_data[fam_data['iid'].isin(sampled_ids)]\n",
    "bim_data = pyp.get_bim()\n",
    "bim_data['gene'] = bim_data['snp'].str.split(':').str[-1]\n",
    "filtered_bim = bim_data[bim_data['gene'].isin(aortopathy_genes)]\n",
    "lof_results = []\n",
    "\n",
    "for marker in filtered_bim['snp']:\n",
    "    genotypes = pyp.get_geno_marker(marker) \n",
    "    for idx, genotype in enumerate(genotypes):\n",
    "        patient_id = fam_data.iloc[idx]['iid']\n",
    "        if genotype in [1, 2]:  # Heterozygous or homozygous var allele\n",
    "            lof_results.append({\n",
    "                \"PMBB_ID\": patient_id,\n",
    "                \"Gene\": filtered_bim.loc[filtered_bim['snp'] == marker, 'gene'].values[0],\n",
    "                \"Marker\": marker,\n",
    "                \"Genotype\": genotype\n",
    "            })\n",
    "\n",
    "lof_results_df = pd.DataFrame(lof_results)\n",
    "output_file = \"C:/bmin520/subgraph/patients_filtered/lof_results.csv\"\n",
    "lof_results_df.to_csv(output_file, index=False)"
   ],
   "id": "e804be6ed31a538e"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "6decc4d094077dd"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
